Edit and redo ConfigMap docs (#35160)
kubernetes/website: content/en/docs/tasks/configure-pod-container/configure-pod-configmap.md

<!-- overview -->
Many applications rely on configuration which is used during either application initialization or runtime.
Most of the times there is a requirement to adjust values assigned to configuration parameters.

> ConfigMaps is the Kubernetes way to inject application pods with configuration data.

Reviewer: Optional extra fix:
```suggestion
ConfigMaps are the Kubernetes way to inject application pods with configuration data.
```

### Restrictions

> - You must create a ConfigMap before referencing it in a Pod specification, or mark the ConfigMap as "optional" (see [Optional ConfigMaps](#optional-configmaps)). If you reference a ConfigMap that doesn't exist, or hasn't been marked as "optional" the Pod won't start. Likewise, references to keys that don't exist in the ConfigMap will prevent the pod from starting.

Reviewer 2: Feels like this makes the sentence harder to understand. 

```suggestion
- You must create the `ConfigMap` object before you reference it in a Pod specification. Alternatively, mark the ConfigMap reference as `optional` in the Pod spec (see [Optional ConfigMaps](#optional-configmaps)). If you reference a ConfigMap that doesn't exist and you don't mark the reference as `optional`, the Pod won't start. Similarly, references to keys that don't exist in the ConfigMap will also prevent the Pod from starting, unless you mark the key references as `optional`.
```

Author: Updated in https://github.com/kubernetes/website/pull/35160/commits/ad33b0c10757978ffaddfa66a691dc9ea302bec1

- You can't use ConfigMaps for {{< glossary_tooltip text="static pods" term_id="static-pod" >}}, because the Kubelet does not support this.

### Optional ConfigMaps

> A ConfigMap reference may be marked "optional".

Reviewer: ```suggestion
In a Pod, or pod template, you can mark a reference to a ConfigMap as _optional_.
```

> If the ConfigMap is non-existent, the mounted volume will be empty.

Reviewer: BTW this isn't quite right because there isn't necessarily any mounted volume.

Author: Good point, I missed that - have expanded this a bit, and modified the wording in the subsection that is specifically about mounted volumes in https://github.com/kubernetes/website/pull/35160/commits/7e16543b9dbe6771017730199b62e2b99c18803e.

If the ConfigMap exists, but the referenced key is non-existent the path will be absent beneath the mount point.

#### Optional ConfigMap in environment variables

There might be situations where environment variables are not always required.

> These environment variables can be marked as optional in a pod like so:

Reviewer: ```suggestion
You can mark an environment variables for a container like this:
```

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "env" ]
      env:
        - name: SPECIAL_LEVEL_KEY
          valueFrom:
            configMapKeyRef:
              name: a-config
              key: akey

>               optional: true

Reviewer: ```suggestion
              optional: true # mark the variable as optional
```

  restartPolicy: Never
```

> When this Pod is run, the output will be empty.

Reviewer: ```suggestion
If you run this pod, and there is no ConfigMap named `a-config`, the output is empty.
If you run this pod, and there is a ConfigMap named `a-config` but that ConfigMap doesn't have
a key named `akey`, the output is also empty. If you do set a value for `akey` in the `a-config`
ConfigMap, this pod prints that value and then terminates.
```

#### Optional ConfigMap via volume plugin

Volumes and files provided by a ConfigMap can be also be marked as optional.
The ConfigMap or the key specified does not have to exist.
The mount path for such items will always be created.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "ls /etc/config" ]
      volumeMounts:
      - name: config-volume
        mountPath: /etc/config
  volumes:
    - name: config-volume
      configMap:
        name: no-config

>         optional: true

Reviewer: ```suggestion
        optional: true # mark the source ConfigMap as optional
```

  restartPolicy: Never
```

> When this pod is run, the output will be:

Reviewer: ```suggestion
If you run this pod, and there is no ConfigMap named `no-config`, the output is:
```

> ```shell

Reviewer: (nit) the output is not shell

Author: Updated in https://github.com/kubernetes/website/pull/35160/commits/7e16543b9dbe6771017730199b62e2b99c18803e

- You can't use ConfigMaps for {{< glossary_tooltip text="static pods" term_id="static-pod" >}}, because the Kubelet does not support this.

### Optional ConfigMaps

> In a Pod, or pod template, you can mark a reference to a ConfigMap as _optional_.

Reviewer 2: ```suggestion
You can mark a reference to a ConfigMap as _optional_ in a Pod specification.
```

Author: Updated in https://github.com/kubernetes/website/pull/35160/commits/ad33b0c10757978ffaddfa66a691dc9ea302bec1

> If the ConfigMap is non-existent, the configuration for which it provides data in the Pod (e.g. environment variable, mounted volume) will be empty.

Reviewer 2: nit
```suggestion
If the ConfigMap doesn't exist, the configuration for which it provides data in the Pod (e.g. environment variable, mounted volume) will be empty.
```

Author: Updated in https://github.com/kubernetes/website/pull/35160/commits/ad33b0c10757978ffaddfa66a691dc9ea302bec1

If the ConfigMap exists, but the referenced key is non-existent the data is also empty.

> #### Optional ConfigMap in environment variables

Reviewer 2: This is just an example of an optional ConfigMap, yes? The content currently makes it look like marking environment variables optional is a special case.

Maybe remove the heading and introduce this as an example:

"For example, the following Pod specification marks an environment variable from a ConfigMap as optional:"

Author: Sure - have updated in https://github.com/kubernetes/website/pull/35160/commits/ad33b0c10757978ffaddfa66a691dc9ea302bec1

There might be situations where environment variables are not always required.
You can mark an environment variables for a container as optional,
like this:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "env" ]
      env:
        - name: SPECIAL_LEVEL_KEY
          valueFrom:
            configMapKeyRef:
              name: a-config
              key: akey
              optional: true # mark the variable as optional
  restartPolicy: Never
```

If you run this pod, and there is no ConfigMap named `a-config`, the output is empty.
If you run this pod, and there is a ConfigMap named `a-config` but that ConfigMap doesn't have
a key named `akey`, the output is also empty. If you do set a value for `akey` in the `a-config`
ConfigMap, this pod prints that value and then terminates.

> #### Optional ConfigMap via volume plugin
> 
> Volumes and files provided by a ConfigMap can be also be marked as optional.
> The ConfigMap or the key specified does not have to exist.
> The mount path for such items will always be created.
> 
> ```yaml
> apiVersion: v1
> kind: Pod
> metadata:
>   name: dapi-test-pod
> spec:
>   containers:
>     - name: test-container
>       image: gcr.io/google_containers/busybox
>       command: [ "/bin/sh", "-c", "ls /etc/config" ]
>       volumeMounts:
>       - name: config-volume
>         mountPath: /etc/config
>   volumes:
>     - name: config-volume
>       configMap:
>         name: no-config
>         optional: true # mark the source ConfigMap as optional
>   restartPolicy: Never
> ```
> 
> If you run this pod, and there is no ConfigMap named `no-config`, the mounted volume will be empty.

Reviewer 2: ```suggestion
You can also mark the volumes and files provided by a ConfigMap as optional. Kubernetes always creates the mount paths for the volume, even if the referenced ConfigMap or key doesn't exist. For example, the following
Pod specification marks a volume that references a ConfigMap as optional:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "ls /etc/config" ]
      volumeMounts:
      - name: config-volume
        mountPath: /etc/config
  volumes:
    - name: config-volume
      configMap:
        name: no-config
        optional: true # mark the source ConfigMap as optional
  restartPolicy: Never
```

If you run this Pod, and there is no ConfigMap named `no-config`, the mounted volume will be empty.
```

Reviewer 2: GitHub messed up the formatting here but I didn't modify the YAML, just the intro

### Mounted ConfigMaps are updated automatically

When a mounted ConfigMap is updated, the projected content is eventually updated too.  This applies in the case where an optionally referenced ConfigMap comes into
existence after a pod has started.

> Kubelet checks whether the mounted ConfigMap is fresh on every periodic sync. However, it uses its local TTL-based cache for getting the current value of the

Reviewer 2: ```suggestion
The kubelet checks whether the mounted ConfigMap is fresh on every periodic sync. However, it uses its local TTL-based cache for getting the current value of the
```

### Optional ConfigMaps

You can mark a reference to a ConfigMap as _optional_ in a Pod specification.

> If the ConfigMap doesn't exist, the configuration for which it provides data in the Pod (e.g. environment variable, mounted volume) will be empty.

Reviewer: A nit: we [avoid Latin phrases](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases) and their abbreviations.
